File: src/main/webapp/static/script/SortParameter.js
class SortParameter {

    static DESC_PREFIX = "-";
    static ASC_PREFIX = "+";

    #field;
    #order;

    constructor(field, order) {
        this.#field = field;
        this.#order = order;
    }

    static fromPrefixedString(str) {
        let field = str;
        let order = SORTING_ORDER.ASC;
        if(str.charAt(0) == SortParameter.DESC_PREFIX) {
            order = SORTING_ORDER.DESC;
            field = str.substring(1, str.length);
        } else if (str.charAt(0) == SortParameter.ASC_PREFIX) {
            field = str.substring(1, str.length);
        }
        return new SortParameter(field, order);
    }

    toPrefixedString() {
        return (this.order == SORTING_ORDER.ASC ? "" : SortParameter.DESC_PREFIX) + this.field;
    }

    get field() {
        return this.#field;
    }

    get order() {
        return this.#order;
    }

}

File: src/main/java/com/nephest/battlenet/sc2/model/web/SortParameter.java
// Copyright (C) 2020-2025 Oleksandr Masniuk
// SPDX-License-Identifier: AGPL-3.0-or-later

package com.nephest.battlenet.sc2.model.web;

import com.nephest.battlenet.sc2.model.SortingOrder;
import jakarta.validation.constraints.NotNull;

public record SortParameter(@NotNull String field, @NotNull SortingOrder order)
{

    public static final String PREFIX_DESC = "-";
    public static final String PREFIX_ASC = "+";
    public static final String SUFFIX_DESC = "desc";
    public static final String SUFFIX_ASC = "asc";
    public static final String SUFFIX_DELIMITER = ":";

    public String toPrefixedString()
    {
        return switch (order)
        {
            case ASC -> field;
            case DESC -> PREFIX_DESC + field;
        };
    }

    public String toSuffixedString()
    {
        return switch (order)
        {
            case ASC -> field + SUFFIX_DELIMITER + SUFFIX_ASC;
            case DESC -> field + SUFFIX_DELIMITER + SUFFIX_DESC;
        };
    }

}
